src/Compiler/Symbols/FSharpDiagnostic.fs

```diff
@@ -9,19 +9,112 @@ namespace FSharp.Compiler.Diagnostics
+        | ContextInfo.OmittedElseBranch _ -> OmittedElseBranch
+        | ContextInfo.ElseBranchResult _ -> ElseBranchResult
+        | ContextInfo.RecordFields -> RecordFields
+        | ContextInfo.TupleInRecordFields -> TupleInRecordFields
+        | ContextInfo.CollectionElement _ -> CollectionElement
+        | ContextInfo.ReturnInComputationExpression -> ReturnInComputationExpression
+        | ContextInfo.YieldInComputationExpression -> YieldInComputationExpression
+        | ContextInfo.RuntimeTypeTest _ -> RuntimeTypeTest
+        | ContextInfo.DowncastUsedInsteadOfUpcast _ -> DowncastUsedInsteadOfUpcast
+        | ContextInfo.FollowingPatternMatchClause _ -> FollowingPatternMatchClause
+        | ContextInfo.PatternMatchGuard _ -> PatternMatchGuard
+        | ContextInfo.SequenceExpression _ -> SequenceExpression
+
+[<Interface>]
+type IFSharpDiagnosticExtendedData = interface end
+
+type TypeMismatchDiagnosticExtendedData
+    internal (symbolEnv: SymbolEnv, dispEnv: DisplayEnv, expectedType: TType, actualType: TType, context: DiagnosticContextInfo) =
+    interface IFSharpDiagnosticExtendedData
+
+    member x.ExpectedType = FSharpType(symbolEnv, expectedType)
+    member x.ActualType = FSharpType(symbolEnv, actualType)
+    member x.ContextInfo = context
+    member x.DisplayContext = FSharpDisplayContext(fun _ -> dispEnv)
+
+type ExpressionIsAFunctionExtendedData
+    internal (symbolEnv: SymbolEnv, actualType: TType) =
+    interface IFSharpDiagnosticExtendedData
+
+    member x.ActualType = FSharpType(symbolEnv, actualType)
+
+type FieldNotContainedDiagnosticExtendedData
+    internal (symbolEnv: SymbolEnv, implTycon: Tycon, sigTycon: Tycon, signatureField: RecdField, implementationField: RecdField) =
+    interface IFSharpDiagnosticExtendedData
+    member x.SignatureField = FSharpField(symbolEnv, RecdFieldRef.RecdFieldRef(mkLocalTyconRef sigTycon, signatureField.Id.idText))
+    member x.ImplementationField = FSharpField(symbolEnv, RecdFieldRef.RecdFieldRef(mkLocalTyconRef implTycon, implementationField.Id.idText))
+
+type ValueNotContainedDiagnosticExtendedData
+    internal (symbolEnv: SymbolEnv, signatureValue: Val, implValue: Val) =
+    interface IFSharpDiagnosticExtendedData
+    member x.SignatureValue = FSharpMemberOrFunctionOrValue(symbolEnv, mkLocalValRef signatureValue)
+    member x.ImplementationValue = FSharpMemberOrFunctionOrValue(symbolEnv, mkLocalValRef implValue)
+
+type ArgumentsInSigAndImplMismatchExtendedData
+    internal(sigArg: Ident, implArg: Ident) =
+    interface IFSharpDiagnosticExtendedData
+    member x.SignatureName = sigArg.idText
+    member x.ImplementationName = implArg.idText
+    member x.SignatureRange = sigArg.idRange
+    member x.ImplementationRange = implArg.idRange
+
+type FSharpDiagnostic(m: range, severity: FSharpDiagnosticSeverity, message: string, subcategory: string, errorNum: int, numberPrefix: string, extendedData: IFSharpDiagnosticExtendedData option) =
```

Comment:
For composability purposes, what about changing "option" to "list" here ?
That way, instead of each diagnostics having its own crafted type, the extended data could compose based on the contents from reusable data parts.